File: answers/lintcode/Best Time to Buy and Sell Stock IV.cpp
class Solution {
public:
    /**
     * @param k: An integer
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    int maxProfit(int k, vector<int> &prices) {
        // write your code here
        if (0 == prices.size()) {
            return 0;
        }
        else if (k >= prices.size()) {
            return maxProfit(prices);
        }
        vector<int> local(k + 1, 0);
        vector<int> global(k + 1, 0);
        for (size_t i = 0; i + 1 < prices.size(); ++ i) {
            int diff = prices[i + 1] - prices[i];
            for (int j = k; j >= 1; -- j) {
                local[j] = max(global[j - 1] + max(diff, 0), local[j] + diff);
                global[j] = max(global[j], local[j]);
            }
        }
        return global[k];
    }
    int maxProfit(const vector<int> &prices) {
        int profit = 0;
        for (size_t i = 1; i < prices.size(); ++ i) {
            if (prices[i] - prices[i - 1] > 0) {
                profit += prices[i] - prices[i - 1];
            }
        }
        return profit;
    }
};

File: answers/leetcode/Best Time to Buy and Sell Stock IV/Best Time to Buy and Sell Stock IV.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        if (k == 0) {
            return 0;
        } else if (k * 2 > prices.length) {
            int sum = 0;
            for (int i = 1; i < prices.length; i++) {
                if (prices[i] > prices[i - 1]) {
                    sum += prices[i] - prices[i - 1];
                }
            }
            return sum;
        } else {
            int[][] dp = new int[k][prices.length];
            for (int i = 0; i < k; i++) {
                int previous = Integer.MIN_VALUE;
                for (int j = 1; j < prices.length; j++) {
                    previous = Math.max(previous, (i > 0 ? dp[i - 1][j - 1] : 0) - prices[j - 1]);
                    dp[i][j] = Math.max(dp[i][j - 1], previous + prices[j]);
                }
            }
            return dp[k - 1][prices.length - 1];
        }
    }
}